### DISCUSSION TITLE
Allowing prefetch outside of component but still client-side?

### DISCUSSION THREAD

**Discussion by corn-flour** [trpc/trpc#3148] [Ideas]
I was trying out the new tanstack router and trying to integrate it with tRPC. This works fine, but I also want to use the loader function of tanstack router to prefetch the data to prevent some waterfall, and trpc currently only expose the `prefetch()` function through the `useContext()` hook, which cannot be called in the loader function since it's not a react component. There's also `createProxySSGHelpers()`, but loader is also not on the server either, so that's not usable for this use case either.

The [with-trpc-react-query](https://github.com/TanStack/router/tree/beta/examples/react/with-trpc-react-query) example on tanstack router's repo just basically ignored trpc completely in the loader functions and define their own fetchData() function for the prefetching, which is definitely not great:
```ts

// loader props of a route whose element uses trpc
loader: async () => {
  queryClient.getQueryData(['posts']) ??
    (await queryClient.prefetchQuery(['posts'], fetchPosts))
  return {}
},

// custom fetchPosts function that doesn't use trpc at all
async function fetchPosts() {
  console.log('Fetching posts...')
  await new Promise((r) => setTimeout(r, 500))
  return axios
    .get<PostType[]>('https://jsonplaceholder.typicode.com/posts')
    .then((r) => r.data.slice(0, 10))
}
```

I tried to avoid having to define a fetch function again by creating a vanilla client and use the query function there instead:
```ts
// create a vanilla trpc client
const client = createTRPCProxyClient<AppRouter>({
  links: [
    httpBatchLink({
      url: "/api/trpc",
    }),
  ],
});

// loader props of route
loader: async () => {
  if (!queryClient.getQueryData(["posts"])) {
    await queryClient.prefetchQuery(["posts"], () => client.posts.query());
  }
  return {};
},
```
This is better in that I don't have to define the fetch function again, but having to call queryClient manually and figuring out the keys that I want to set is still not ideal I think. If possible, I think something like this is probably better:

```ts
loader: async () => {
  if (!queryClients.posts.getQueryData()) {
    await queryClients.posts.prefetchQuery();
  }
  return {};
},
```
Thank you so much!

**Comment by jamesopstad** [comment]
I am having similar issues trying to use tRPC with React Router 6.4. When using React Router together with React Query it is recommended to call the query client's methods directly inside loaders and actions (https://reactrouter.com/en/main/guides/data-libs). However, tRPC only exposes these methods via the `useContext` hook and this can't be accessed outside React components. It would be great if the query client could be accessed outside components, similar to how the vanilla client can. I think the solution would be for `client` and `queryClient` to be combined before being passed to `trpc.Provider`, possibly by providing the `queryClient` as an argument to `trpc.createClient`. What do you think?

**Comment by KATT** [comment]
If you're doing an SPA, you can just create a global tRPC object & query client

Dummy-code (not tested):


⚠️⚠️ ⚠️  Don't do this if you're doing SSR as it will reuse the same query client across **all** users.

Based on https://trpc.io/docs/react
```ts
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { httpBatchLink } from '@trpc/client';
import React, { useState } from 'react';
import { trpc } from './utils/trpc';

export const queryClient = new QueryClient();
export const trpcClient = new trpc.createClient({
  links: [
    httpBatchLink({
      url: 'http://localhost:5000/trpc',
      // optional
      headers() {
        return {
          authorization: getAuthCookie(),
        };
      },
    }),
  ],
});

let trpcUtils: ReturnType<typeof trpc['useContext']> | null = null;
export function getTRPCUtils() {
  if (!trpcUtils) {
    throw new Error("Something went wrong")
  }
  
  return null;
}


function GlobalUtils() {
  const context = trpc.useContext();
  trpcUtils = context;
  return null;
}

export function App() {
  return (
    <trpc.Provider client={trpcClient} queryClient={queryClient}>
      <QueryClientProvider client={queryClient}>
        <GlobalUtils />
        {/* Your app here */}
      </QueryClientProvider>
    </trpc.Provider>
  );
}
```

**Reply by jamesopstad** [reply]
Thanks for looking into this. Unfortunately this approach won't work because the React Router loaders are executed before the React tree has resolved. This means `trpcUtils` is still `null`. I think the only way to get this working is for the context to be created outside of React.

**Comment by abetoots** [comment]
@jamesopstad @bte234 There's an example in this [blog post](https://tkdodo.eu/blog/react-query-meets-react-router). I think this could be combined with `getQueryKey` as defined [here](https://trpc.io/docs/getQueryKey). 

Edit: So I've got the logic down up until getting the already fetched data.
```
export const loader = async (queryClient: QueryClient) =>  (args: LoaderFunctionArgs) =>{
    const queryKey = getQueryKey(trpc.dashboard.get)
    //get already fetched data or prefetch it using trpc
    return queryClient.getQueryData(queryKey) ?? //prefetch here
}
```
Problem is I don't know how to start pre fetching the TRPC query.
